src/hotspot/share/prims/jvmtiExport.cpp

```diff
@@ -125,11 +131,18 @@ class JvmtiThreadEventTransition : StackObj {
   }
 
   ~JvmtiThreadEventTransition() {
-    if (_jthread != nullptr)
+    if (_jthread != nullptr) {
       ThreadStateTransition::transition_from_native(_jthread, _saved_state);
+    }
+    JvmtiEventController::dec_in_callback_count();
   }
 };
 
+// The JVMTI_...__BLOCK are used to ensure that vm_death is the last posted event.
+// The callbacks are not executed after _execution_finished is set to true
+// and the _in_callback_count contains the number of callbacks still in progress.
+#define JVMTI_JAVA_EVENT_CALLBACK_BLOCK(thread) JvmtiJavaThreadEventTransition jet(thread); if (JvmtiEventController::is_execution_finished()) { return; }
+#define JVMTI_EVENT_CALLBACK_BLOCK(thread) JvmtiThreadEventTransition jet(thread); if (JvmtiEventController::is_execution_finished()) { return; }
```

Comment:
Nit:
It is better to keep macro names congruent with matching classes:
  s/JVMTI_JAVA_EVENT_CALLBACK_BLOCK/JVMTI_JAVA_THREAD_EVENT_CALLBACK_BLOCK/g
  s/JVMTI_EVENT_CALLBACK_BLOCK/JVMTI_THREAD_EVENT_CALLBACK_BLOCK/g
    
Also, I'd suggest to use slashes in these macro definitions:
```
#define JVMTI_JAVA_THREAD_EVENT_CALLBACK_BLOCK(thread) \
  JvmtiJavaThreadEventTransition jet(thread); \
  if (JvmtiEventController::is_execution_finished()) { \
    return; \
  }
```